### PR TITLE
Add withMap to MutableMap

### PR DESCRIPTION
**Pull request by aboullaite**
Signed-off-by: Mohammed Aboullaite <aboullaite.mohammed@gmail.com>

This PR adds `withMap` a requested in #761

### CODE REVIEW

File: .idea/checkstyle-idea.xml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (nikhilnanivadekar) [inline .idea/checkstyle-idea.xml]: Please revert this file

File: unit-tests/src/test/java/org/eclipse/collections/impl/map/mutable/MutableMapIterableTestCase.java [hunk 2]
```diff
@@ -560,6 +563,18 @@ public void withKeyValue()
         Verify.assertMapsEqual(UnifiedMap.newWithKeysValues("A", 11, "B", 2), mapWith);
     }
 
+    @Test
+    public void withMap()
+    {
+        MutableMapIterable<String, Integer> map = this.newMapWithKeysValues("A", 1, "B", 2);
```
Comment (nikhilnanivadekar) [inline unit-tests/src/test/java/org/eclipse/collections/impl/map/mutable/MutableMapIterableTestCase.java]: Can you add a test case for empty maps?
1) Input map is empty, target map is non-empty
2) Input map is empty, target map is empty
3) Input map is non-empty, target map is empty

**Comment by nikhilnanivadekar** [comment]
@aboullaite thanks for your contribution.

1. Your Eclipse Foundation ECA validation is not passing. Please refer to this to follow the instructions for getting it to pass:
2. I asked for a few more tests in the review comments. These are for regression.
3. The builds are failing due to test failures.

**Comment by aboullaite** [comment]
Thanks @nikhilnanivadekar for your feedback! 1 and 2 are fixed! I am trying to get the build pass right now.
I have a question, now with `withMap` if the argument is `null`, this will trigger a `NullPointerException`. Is it the intended behavior ? should I check for nullability and trigger `IllegalArgumentException` instead.  also should I add test about this specific case ?

**Comment by nikhilnanivadekar** [comment]
Not supporting nulls is an intended behavior but it’ll be great if you could throw IllegalArgumentException instead of an NPE

**Comment by motlin** [comment]
NPE is also fine for null. It's nice to use Objects.requireNonNull with a
helpful error message.

**Comment by nikhilnanivadekar** [comment]
But isn’t IAE better than NPE for tracing and maintenance? Asking honestly, not a rhetorical question

File: unit-tests/src/test/java/org/eclipse/collections/impl/map/mutable/MutableMapIterableTestCase.java (hunk 2 above)
Comment (aboullaite) [reply unit-tests/src/test/java/org/eclipse/collections/impl/map/mutable/MutableMapIterableTestCase.java]: I added them!

**Comment by aboullaite** [comment]
I've opted for IAE for the moment! If we decide to switch to a NPE with a `requireNonNull` I can do that.

WDYT?

**Comment by nikhilnanivadekar** [comment]
sounds good!

**Comment by motlin** [comment]
It's not a big deal either way, but Effective Java Item 72: _Favor the use of standard exceptions_ says "If a caller passes `null` in some parameter for which null values are prohibited, convention dictates that `NullPointerException` be thrown rather than `IllegalArgumentException`."

**Comment by nikhilnanivadekar** [comment]
@aboullaite we are close to completing it. The Javadoc needs a little bit of editing. The implementation is correct, in the tests, I have suggested simpler and Eclipse Collections ways. Please reach out if you have any questions.

**Comment by aboullaite** [comment]
@nikhilnanivadekar I'll give the javadoc a second read! However, AFAIK I've implemented the tests you've suggested! Am I missing something ?

**Comment by nikhilnanivadekar** [comment]
Tests are good and correct. Just a few stylistic comments to make the code condensed and use Eclipse Collections :)

**Comment by aboullaite** [comment]
I modified the javadoc for `MutableMapIterable.java`. However could you elaborate more on:
> Tests are good and correct. Just a few stylistic comments to make the code condensed and use Eclipse Collections :)

Thanks

**Comment by nikhilnanivadekar** [comment]
I have already added line comments. I am away from my machine right now, but I can push the changes to test to my branch for reference. Please let me know.

File: unit-tests/src/test/java/org/eclipse/collections/impl/map/mutable/MutableMapIterableTestCase.java [hunk 3]
```diff
@@ -560,6 +563,50 @@ public void withKeyValue()
...
+    }
+
+    @Test
+    public void withMapTargetEmpty()
+    {
+        MutableMapIterable<String, Integer> map = this.newMap();
+        Map<String, Integer> simpleMap = Stream.of(new Object[][] {
+                { "A", 1 },
+                { "B", 2 },
+        }).collect(Collectors.toMap(data -> (String) data[0], data -> (Integer) data[1]));
```
Comment (nikhilnanivadekar) [inline unit-tests/src/test/java/org/eclipse/collections/impl/map/mutable/MutableMapIterableTestCase.java]:
```suggestion
        Map<String, Integer> simpleMap = Maps.mutable.with("A", 1, "B", 2);
```

File: unit-tests/src/test/java/org/eclipse/collections/impl/map/mutable/MutableMapIterableTestCase.java [hunk 4]
```diff
@@ -560,6 +563,50 @@ public void withKeyValue()
...
+        MutableMapIterable<String, Integer> mapWith = map.withMap(simpleMap);
+        Assert.assertSame(map, mapWith);
+        Verify.assertMapsEqual(UnifiedMap.newWithKeysValues("A", 1, "B", 2), mapWith);
+    }
+
+    @Test
+    public void withMapEmptyAndTargetEmpty()
+    {
+        MutableMapIterable<String, Integer> map = this.newMap();
+        MutableMapIterable<String, Integer> mapWith = map.withMap(Collections.emptyMap());
```
[the remark this answers is not in the document]
Comment (nikhilnanivadekar) [reply unit-tests/src/test/java/org/eclipse/collections/impl/map/mutable/MutableMapIterableTestCase.java]:
```suggestion
        MutableMapIterable<String, Integer> mapWith = map.withMap(Maps.mutable.empty());
```

File: unit-tests/src/test/java/org/eclipse/collections/impl/map/mutable/MutableMapIterableTestCase.java [hunk 5]
```diff
@@ -638,6 +685,7 @@ public void rehash_null_collision()
     public void updateValue()
     {
         MutableMapIterable<Integer, Integer> map = this.newMap();
+
```
[the remark this answers is not in the document]
Comment (nikhilnanivadekar) [reply unit-tests/src/test/java/org/eclipse/collections/impl/map/mutable/MutableMapIterableTestCase.java]:
```suggestion
```

File: unit-tests/src/test/java/org/eclipse/collections/impl/map/mutable/UnmodifiableMutableMapTest.java [hunk 6]
```diff
@@ -284,6 +288,36 @@ public void withKeyValue()
         this.newMapWithKeysValues(1, "One", 2, "2").withKeyValue(null, null);
     }
 
+    @Override
+    @Test
+    public void withMap()
+    {
+        Verify.assertThrows(UnsupportedOperationException.class, () -> this.newMapWithKeyValue(1, 'a').withMap(
+                Stream.of(new Object[][] {{ 1, 'a' }}).collect(Collectors.toMap(data -> (Integer) data[0], data -> (Character) data[1]))));
```
Comment (nikhilnanivadekar) [inline unit-tests/src/test/java/org/eclipse/collections/impl/map/mutable/UnmodifiableMutableMapTest.java]:
```suggestion
        Verify.assertThrows(UnsupportedOperationException.class, () -> this.newMapWithKeyValue(1, 'a').withMap(
                Maps.mutable.with(1, Character.valueOf("a"))));
```

File: unit-tests/src/test/java/org/eclipse/collections/impl/map/mutable/UnmodifiableMutableMapTest.java [hunk 7]
```diff
@@ -284,6 +288,36 @@ public void withKeyValue()
...
+    }
+
+    @Override
+    @Test
+    public void withMapEmpty()
+    {
+        Verify.assertThrows(UnsupportedOperationException.class, () -> this.newMapWithKeyValue(1, 'a').withMap(Collections.emptyMap()));
```
Comment (nikhilnanivadekar) [inline unit-tests/src/test/java/org/eclipse/collections/impl/map/mutable/UnmodifiableMutableMapTest.java]:
```suggestion
        Verify.assertThrows(UnsupportedOperationException.class, () -> this.newMapWithKeyValue(1, 'a').withMap(Maps.mutable.empty()));
```

File: unit-tests/src/test/java/org/eclipse/collections/impl/map/mutable/UnmodifiableMutableMapTest.java [hunk 8]
```diff
@@ -284,6 +288,36 @@ public void withKeyValue()
...
+    }
+
+    @Override
+    @Test
+    public void withMapTargetEmpty()
+    {
+        Verify.assertThrows(UnsupportedOperationException.class, () -> this.newMap().withMap(
+                Stream.of(new Object[][] {{ 1, 'a' }}).collect(Collectors.toMap(data -> (Integer) data[0], data -> (Character) data[1]))));
```
Comment (nikhilnanivadekar) [inline unit-tests/src/test/java/org/eclipse/collections/impl/map/mutable/UnmodifiableMutableMapTest.java]:
```suggestion
        Verify.assertThrows(UnsupportedOperationException.class, () -> this.newMap().withMap(Maps.mutable.with(1, Character.valueOf("a")));
```

File: unit-tests/src/test/java/org/eclipse/collections/impl/map/mutable/UnmodifiableMutableMapTest.java [hunk 9]
```diff
@@ -284,6 +288,36 @@ public void withKeyValue()
...
+    }
+
+    @Override
+    @Test
+    public void withMapEmptyAndTargetEmpty()
+    {
+        Verify.assertThrows(UnsupportedOperationException.class, () -> this.newMap().withMap(Collections.emptyMap()));
```
Comment (nikhilnanivadekar) [inline unit-tests/src/test/java/org/eclipse/collections/impl/map/mutable/UnmodifiableMutableMapTest.java]:
```suggestion
        Verify.assertThrows(UnsupportedOperationException.class, () -> this.newMap().withMap(Maps.mutable.empty()));
```

File: unit-tests/src/test/java/org/eclipse/collections/impl/map/sorted/mutable/UnmodifiableTreeMapTest.java [hunk 10]
```diff
@@ -348,6 +351,36 @@ public void withKeyValue()
         this.newMapWithKeysValues(1, "One", 2, "2").withKeyValue(null, null);
     }
 
+    @Override
+    @Test
+    public void withMap()
+    {
+        Verify.assertThrows(UnsupportedOperationException.class, () -> this.newMapWithKeyValue(1, 'a').withMap(
+                Stream.of(new Object[][] {{ 1, 'a' }}).collect(Collectors.toMap(data -> (Integer) data[0], data -> (Character) data[1]))));
```
Comment (nikhilnanivadekar) [inline unit-tests/src/test/java/org/eclipse/collections/impl/map/sorted/mutable/UnmodifiableTreeMapTest.java]:
```suggestion
        Verify.assertThrows(UnsupportedOperationException.class, () -> this.newMapWithKeyValue(1, 'a').withMap(Maps.mutable.with(1, Character.valueOf("a"))));
```

File: unit-tests/src/test/java/org/eclipse/collections/impl/map/sorted/mutable/UnmodifiableTreeMapTest.java [hunk 11]
```diff
@@ -348,6 +351,36 @@ public void withKeyValue()
...
+    }
+
+    @Override
+    @Test
+    public void withMapEmpty()
+    {
+        Verify.assertThrows(UnsupportedOperationException.class, () -> this.newMapWithKeyValue(1, 'a').withMap(Collections.emptyMap()));
```
Comment (nikhilnanivadekar) [inline unit-tests/src/test/java/org/eclipse/collections/impl/map/sorted/mutable/UnmodifiableTreeMapTest.java]:
```suggestion
        Verify.assertThrows(UnsupportedOperationException.class, () -> this.newMapWithKeyValue(1, 'a').withMap(Maps.mutable.empty()));
```

File: unit-tests/src/test/java/org/eclipse/collections/impl/map/sorted/mutable/UnmodifiableTreeMapTest.java [hunk 12]
```diff
@@ -348,6 +351,36 @@ public void withKeyValue()
...
+    }
+
+    @Override
+    @Test
+    public void withMapTargetEmpty()
+    {
+        Verify.assertThrows(UnsupportedOperationException.class, () -> this.newMap().withMap(
+                Stream.of(new Object[][] {{ 1, 'a' }}).collect(Collectors.toMap(data -> (Integer) data[0], data -> (Character) data[1]))));
```
Comment (nikhilnanivadekar) [inline unit-tests/src/test/java/org/eclipse/collections/impl/map/sorted/mutable/UnmodifiableTreeMapTest.java]:
```suggestion
        Verify.assertThrows(UnsupportedOperationException.class, () -> this.newMap().withMap(Maps.mutable.with(1, Character.valueOf("a")));
```

File: unit-tests/src/test/java/org/eclipse/collections/impl/map/sorted/mutable/UnmodifiableTreeMapTest.java [hunk 13]
```diff
@@ -348,6 +351,36 @@ public void withKeyValue()
...
+    }
+
+    @Override
+    @Test
+    public void withMapEmptyAndTargetEmpty()
+    {
+        Verify.assertThrows(UnsupportedOperationException.class, () -> this.newMap().withMap(Collections.emptyMap()));
```
Comment (nikhilnanivadekar) [inline unit-tests/src/test/java/org/eclipse/collections/impl/map/sorted/mutable/UnmodifiableTreeMapTest.java]:
```suggestion
        Verify.assertThrows(UnsupportedOperationException.class, () -> this.newMap().withMap(Maps.mutable.empty()));
```

**Comment by nikhilnanivadekar** [comment]
I forgot that I should give code suggestions on github itself. I have done that in the tests. Also, I apologize, I had forgotten to submit my previous review so it was sitting in draft forever. I have submitted it now. Hopefully the review makes it clear. 
Also, this is a great way for you to see the difference between a concise way of writing code using Eclipse Collections compared to the verbose Java Collections way.
Lastly, for the Javadoc, I dont think we are supported fixedsize and immutable maps. So, you can remove those terms.
This is almost close to a merge! Thanks for your help and contribution! 👍

File: eclipse-collections-api/src/main/java/org/eclipse/collections/api/map/MutableMapIterable.java [hunk 14]
```diff
@@ -152,6 +152,25 @@ default boolean removeIf(Predicate2<? super K, ? super V> predicate)
      */
     MutableMapIterable<K, V> withKeyValue(K key, V value);
 
+    /**
+     * This method allows mutable, fixed size, and immutable maps the ability to add elements of a Map to their existing
```
[the remark this answers is not in the document]
Comment (aboullaite) [reply eclipse-collections-api/src/main/java/org/eclipse/collections/api/map/MutableMapIterable.java]: I am planning to! but I'll fix this one first, merge, and work on fixed size and immutable maps support as well.

**Comment by aboullaite** [comment]
Thanks a lot @nikhilnanivadekar for your patience and support! I've made a new commit solving all the improvements you've mentioned :)

File: eclipse-collections-api/src/main/java/org/eclipse/collections/api/map/MutableMapIterable.java [hunk 15]
```diff
@@ -152,6 +152,25 @@ default boolean removeIf(Predicate2<? super K, ? super V> predicate)
...
+     * map = map.withKeyValue("new key", "new value");
+     * </pre>
+     * In the case of FixedSizeMap, a new instance will be returned by withKeyValue, and any variables that
+     * previously referenced the original map will need to be redirected to reference the new instance. In the case
+     * of a FastMap or UnifiedMap, you will be replacing the reference to map with map, since FastMap and UnifiedMap
+     * will both return "this" after calling put on themselves.
+     *
+     * @see #put(Object, Object)
+     */
+    MutableMapIterable<K, V> withMap(Map<K, V> map);
```
[the remark this answers is not in the document]
Comment (motlin) [reply eclipse-collections-api/src/main/java/org/eclipse/collections/api/map/MutableMapIterable.java]: "with mutable maps"

I feel like this must mean "immutable" no?

Does the method return `this`? It should say so.

This behaves kind of like putAll() right? The link in @see should be to putAll if so.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/AbstractMutableBiMap.java [hunk 16]
```diff
@@ -153,6 +153,17 @@ else if (other == value || value.equals(other))
...
+    @Override
+    public MutableBiMap<K, V> withMap(Map<K, V> map)
+    {
+        if (map == null)
+        {
+            throw new IllegalArgumentException("Cannot Add a null Map!");
+        }
+        map.forEach(this::put);
+        return this;
+    }
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/AbstractMutableBiMap.java]: I don't think we should have a null check.

Instead of delegating to forEach, I think we should be delegating to putAll() which sometimes is better optimized (See UnifiedMap for an example).

Taken together, the implementation for all mutable maps will be:
```java
this.putAll(map);
return this;
```
Which means the implementations can be pulled up to interfaces as default methods. AbstractMutable*Maps will be unchanged.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/SynchronizedBiMap.java [hunk 17]
```diff
@@ -291,6 +292,20 @@ public V forcePut(K key, V value)
...
+    {
+        if (map == null)
+        {
+            throw new IllegalArgumentException("Cannot Add a null Map!");
+        }
+        synchronized (this.lock)
+        {
+            map.forEach(this::put);
+            return this;
+        }
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/SynchronizedBiMap.java]: Synchronized wrappers should delegate to methods of the same name. But then return `this`, the wrapper.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/AbstractMutableBiMap.java (hunk 16 above)
Comment (aboullaite) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/AbstractMutableBiMap.java]: The null check is following the discussion we had previously in this PR! is just to trigger IAE instead of NPE! If the default behaviour should be to trigger NPEs instead, then I remove it and update the tests! WDYT ?

Comment (motlin) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/AbstractMutableBiMap.java]: Yes, I replied. Repeating here:

> It's not a big deal either way, but Effective Java Item 72: Favor the use of standard exceptions says "If a caller passes null in some parameter for which null values are prohibited, convention dictates that NullPointerException be thrown rather than IllegalArgumentException."

In addition, this method feels very similar to putAll(). If you call Map.putAll(null) on any map, including our own implementations, you'll get a NullPointerException. None of them have a null check either.

Comment (aboullaite) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/AbstractMutableBiMap.java]: Got it! Thanks

Comment (aboullaite) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/AbstractMutableBiMap.java]: @motlin I'm facing some test failures after removing the null check from the method! some classes (`org.eclipse.collections.impl.map.mutable.ConcurrentHashMap` for example) are triggering IAE since they use `java.util.Map#putAll` method, however others (`org.eclipse.collections.impl.bimap.mutable.HashBiMap`) trigger NPE, since `putAll` method is overridden in `org.eclipse.collections.impl.bimap.mutable.AbstractMutableBiMap`

**Comment by nikhilnanivadekar** [comment]
While being pedantic and following Effective Java is great! But as a user, I feel that NPEs are super unintuitive and hard to narrow down to find the issue. I would rather have an IAE where I get to know what exactly is incorrect.

**Comment by motlin** [comment]
I somewhat agree. This is a method with a single argument. The only way it
can throw NPE is if it's one argument is null. In this case, it's also just
like putAll(), should be implemented in terms of putAll(), and the
exception should even include putAll() in the stack trace.

**Comment by nikhilnanivadekar** [comment]
I think we need to provide direction to @aboullaite about this. I don’t quite understand your line of thought of having putAll() in the stacktrace. 
Let’s get this PR merged by throwing an IAE and delegating to putAll() in the implementation. If there are additional improvements we can do it as a part of another PR.

**Comment by motlin** [comment]
The implementation should be exactly:
```
this.putAll(arg);
return this;
```
and should be a default method on the interface. I would like to be
convinced why it should be something more complex than that.

**Comment by nikhilnanivadekar** [comment]
Now I understand. @aboullaite can you please change the implementation to be only what Craig pointed out? If there are inconsistencies in the exceptions thrown then you’ll need to change the tests for this method accordingly. Thanks for your contribution @aboullaite and thanks @motlin for the clarification!

**Comment by aboullaite** [comment]
Thank you both for your help! On it :)

File: eclipse-collections-api/src/main/java/org/eclipse/collections/api/map/MutableMapIterable.java [hunk 18]
```diff
@@ -161,11 +161,15 @@ default boolean removeIf(Predicate2<? super K, ? super V> predicate)
      * map = map.withMap(someMap);
      * </pre>
      *
-     * @see #put(Object, Object)
+     * @see #putAll(Map)
      *
      * @since 10.3.0
      */
-    MutableMapIterable<K, V> withMap(Map<K, V> map);
+    default MutableMapIterable<K, V> withMap(Map<K, V> map)
```
Comment (motlin) [inline eclipse-collections-api/src/main/java/org/eclipse/collections/api/map/MutableMapIterable.java]: While I do think we should pull the method up, we need it to have the correct generics, which is a breaking change. It needs the bounded wildcards ? extends K, ? extends V

Comment (aboullaite) [reply eclipse-collections-api/src/main/java/org/eclipse/collections/api/map/MutableMapIterable.java]: Oh yes! That is true. Fixing it right away

**Comment by aboullaite** [comment]
@nikhilnaniv @motlin any required action from my side?! Just double checking 😊

**Comment by nikhilnanivadekar** [comment]
@motlin can you check and approve. this looks good from my perspective.

**Comment by nikhilnanivadekar** [comment]
@aboullaite can you squash all commits and rebase. It will be easier to review.

File: eclipse-collections-api/src/main/java/org/eclipse/collections/api/factory/map/MutableMapFactory.java [hunk 19]
```diff
@@ -72,6 +72,9 @@
 
     <K, V> MutableMap<K, V> ofMap(Map<? extends K, ? extends V> map);
 
+    /**
+     * @since 10.3.0
+     */
```
Comment (motlin) [inline eclipse-collections-api/src/main/java/org/eclipse/collections/api/factory/map/MutableMapFactory.java]: Not accurate. @nikhilnanivadekar added this in 2018.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/SynchronizedBiMap.java [hunk 20]
```diff
@@ -291,6 +292,16 @@ public V forcePut(K key, V value)
         }
     }
 
+    @Override
+    public MutableBiMap<K, V> withMap(Map<? extends K, ? extends V> map)
+    {
+        synchronized (this.lock)
+        {
+            this.getDelegate().putAll(map);
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/SynchronizedBiMap.java]: this could be this.putAll instead of this.getDelegate().putAll but I'm not sure if that matters

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/map/mutable/ConcurrentHashMap.java [hunk 21]
```diff
@@ -1979,6 +1979,12 @@ public void zeroOutQueuePosition()
         return (ConcurrentHashMap<K, V>) super.withKeyValue(key, value);
     }
 
+    @Override
+    public ConcurrentHashMap<K, V> withMap(Map<? extends  K, ? extends V> map)
+    {
+        return (ConcurrentHashMap<K, V>) super.withMap(map);
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/map/mutable/ConcurrentHashMap.java]: Any reason not to do this.putAll(); return this; ?

Comment (aboullaite) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/map/mutable/ConcurrentHashMap.java]: I basically followed the same coding style here for all `with*` methods! only reason is consistency

File: eclipse-collections-api/src/main/java/org/eclipse/collections/api/factory/map/MutableMapFactory.java (hunk 19 above)
Comment (aboullaite) [reply eclipse-collections-api/src/main/java/org/eclipse/collections/api/factory/map/MutableMapFactory.java]: My bad! shouldn't be here! fixing it ...

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/map/mutable/ConcurrentHashMap.java (hunk 21 above)
Comment (donraab) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/map/mutable/ConcurrentHashMap.java:1985]: @molin Can you take a look at this and review/approve? I took a quick look and didn't see any casts.

File: eclipse-collections-api/src/main/java/org/eclipse/collections/api/map/MutableMapIterable.java [hunk 22]
```diff
@@ -152,6 +152,25 @@ default boolean removeIf(Predicate2<? super K, ? super V> predicate)
...
     MutableMapIterable<K, V> withKeyValue(K key, V value);
 
+    /**
+     * This method allows mutable maps the ability to add elements of a Map to their existing
+     * elements. The original map is modified and then returned. In order to use this method properly with mutable maps
+     * the following approach must be taken:
+     * <p>
+     * <pre>
+     * map = map.withMap(someMap);
+     * </pre>
```
Comment (motlin) [inline eclipse-collections-api/src/main/java/org/eclipse/collections/api/map/MutableMapIterable.java:162]: The javadoc seems wrong. At first I thought this was going to be about a method on mutable/immutable maps. If you're not sure if the method is mutating or not, then the safe thing to do is:
```
newReference = oldReference.possiblyMutatingMethod(args)
```
This one is specifically on mutable maps so the guidance doesn't need to be, and should not be followed. If you just delete "In order to use this method properly" until the end, I think it's good.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/map/sorted/mutable/SynchronizedSortedMap.java [hunk 23]
```diff
@@ -122,6 +123,16 @@ public SynchronizedSortedMap(MutableSortedMap<K, V> newMap, Object lock)
         }
     }
 
+    @Override
+    public MutableSortedMap<K, V> withMap(Map<? extends K, ? extends V> map)
+    {
+        synchronized (this.lock)
+        {
+            this.putAll(map);
+            return this;
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/map/sorted/mutable/SynchronizedSortedMap.java:132]: Some of the synchronized wrappers call this.putAll and some call this.getDelegate.putAll. I prefer this.putAll, like we see here.

File: unit-tests/src/test/java/org/eclipse/collections/impl/bimap/mutable/UnmodifiableBiMapTest.java [hunk 24]
```diff
@@ -161,6 +164,36 @@ public void withKeyValue()
         Verify.assertThrows(UnsupportedOperationException.class, () -> this.getEmptyMap().withKeyValue(1, 'a'));
     }
 
+    @Override
+    @Test
+    public void withMap()
+    {
+        Verify.assertThrows(UnsupportedOperationException.class, () -> this.newMapWithKeyValue(1, 'a').withMap(
```
[the remark this answers is not in the document]
Comment (motlin) [reply unit-tests/src/test/java/org/eclipse/collections/impl/bimap/mutable/UnmodifiableBiMapTest.java:171]: Instead of "mutation" I think you mean "mutable"

[the remark this answers is not in the document]
Comment (motlin) [reply unit-tests/src/test/java/org/eclipse/collections/impl/bimap/mutable/UnmodifiableBiMapTest.java:171]: edited inline

File: eclipse-collections-api/src/main/java/org/eclipse/collections/api/map/MutableMapIterable.java (hunk 22 above)
Comment (aboullaite) [reply eclipse-collections-api/src/main/java/org/eclipse/collections/api/map/MutableMapIterable.java:162]: Roger that!

**Comment by aboullaite** [comment]
Thanks @motlin, I made a new commit with the requested changes (I didn't notice that @nikhilnanivadekar requested changes as well until you mentioned it :/)

**Comment by nikhilnanivadekar** [comment]
@aboullaite this looks good to me. @motlin you still have any changes required or can you approve the Review as well.
Lastly @aboullaite , a little bit of housekeeping. Can you please change the commit message to have only "Add withMap() to MutableMap. Closes #761 " Please? It helps during preparing release notes. Thanks!

File: eclipse-collections-api/src/main/java/org/eclipse/collections/api/map/MutableMapIterable.java [hunk 25]
```diff
@@ -152,6 +152,20 @@ default boolean removeIf(Predicate2<? super K, ? super V> predicate)
      */
     MutableMapIterable<K, V> withKeyValue(K key, V value);
 
+    /**
+     * This method allows mutable maps the ability to add elements of a Map to their existing
+     * elements. The original map is modified and then returned.
```
Comment (motlin) [inline eclipse-collections-api/src/main/java/org/eclipse/collections/api/map/MutableMapIterable.java:157]: The wording is confusing to me. I'd suggest something like "Copies all of the mappings from the specified map to this map and returns this map" or "Similar to putAll() but returns `this` instead of `void`"

**Comment by aboullaite** [comment]
@motlin I pushed new changes based on your feedback. Can you check ?

**Comment by motlin** [comment]
Looks good now. Just needs to be rebased on top of upstream/master.

**Comment by motlin** [comment]
This was a tricky one. Thanks for sticking with me and getting it landed!
